drivers/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,337 @@
+			      off_t offset,
+			      size_t size);
+static int btctlr_flash_write_protection_set(struct device *dev,
+					     bool enable);
+
+static int flash_op_execute(void);
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void btctlr_flash_page_layout_get(
+	struct device *dev,
+	const struct flash_pages_layout **layout,
+	size_t *layout_size);
+#endif /* defined(CONFIG_FLASH_PAGE_LAYOUT) */
+
+static const struct flash_driver_api btctrl_flash_api = {
+	.read = btctlr_flash_read,
+	.write = btctlr_flash_write,
+	.erase = btctlr_flash_erase,
+	.write_protection = btctlr_flash_write_protection_set,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = btctlr_flash_page_layout_get,
+#endif  /* CONFIG_FLASH_PAGE_LAYOUT */
+	.write_block_size = FLASH_DRIVER_WRITE_BLOCK_SIZE
+};
+
+
+/* Utility functions */
+static inline bool is_addr_valid(off_t addr, size_t len)
+{
+	if ((addr + len) > NRF_FICR->CODEPAGESIZE * NRF_FICR->CODESIZE ||
+	    addr < 0) {
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool is_aligned_32(off_t addr)
+{
+	return ((addr & 0x3) == 0);
+}
+
+static inline off_t align_32(off_t addr)
+{
+	return (addr & ~0x3);
+}
+
+static inline size_t bytes_to_words(size_t bytes)
+{
+	return bytes / sizeof(u32_t);
+}
+
+static inline bool is_page_aligned(off_t addr)
+{
+	return (addr & (NRF_FICR->CODEPAGESIZE - 1)) == 0;
+}
+
+static void flash_operation_complete_callback(u32_t status)
+{
+	__ASSERT_NO_MSG(flash_state.op == FLASH_OP_WRITE ||
+			flash_state.op == FLASH_OP_ERASE);
+
+	int err;
+
+	flash_state.addr += flash_state.prev_len;
+	flash_state.data = (const void *) ((intptr_t) flash_state.data
+					   + flash_state.prev_len);
+	flash_state.len -= flash_state.prev_len;
+
+	if (flash_state.len > 0) {
+		err = flash_op_execute();
+		/* All inputs should have been validated on the first call. */
+		__ASSERT(err == 0, "Continued flash operation failed\n");
+	} else {
+		flash_state.op = FLASH_OP_NONE;
+		k_sem_give(&flash_state.sem);
+	}
+}
+
+static size_t offset_32(const void *addr)
+{
+	return (const u32_t) addr & 0x3;
+}
+
+static size_t unaligned_word_copy(u32_t *word_dst,
+				  void *dst,
+				  const void *src,
+				  size_t len)
+{
+	/* Example: Writing one byte from src to dst via word_dst.
+	 *
+	 * |    |<----- src (len = 1)
+	 * |  |<------- dst
+	 * | 0 1 2 3 |
+	 * | address |
+	 *
+	 * max_offset             : max(1, 2) => 2
+	 * remaining_bytes_in_word: 4 - 2     => 2
+	 * bytes_to_copy          : MIN(1, 2) => 1
+	 */
+	size_t max_offset = MAX(offset_32(dst), offset_32(src));
+	size_t remaining_bytes_in_word = sizeof(u32_t) - max_offset;
+	size_t bytes_to_copy = MIN(len, remaining_bytes_in_word);
+
+	/* nRF52832 Product specification:
+	 *  Only full 32-bit words can be written to Flash using the
+	 *  NVMC interface. To write less than 32 bits to Flash, write
+	 *  the data as a word, and set all the bits that should remain
+	 *  unchanged in the word to '1'.
+	 */
+	*word_dst = ~0;
```

Comment:
What about if the end is unaligned? Don't we need to set the remaining bytes in the word to '1's? Or is that not supported?

Reply:
This is supported. We're inverting the 0: `~0 = 0xFFFFFFFF`. The `memcpy()` does the magic copying and it should handle unaligned start and unaligned end.

Reply:
What I don't understand is, you are only setting 4 bytes to F's. But you might have to write eight bytes.

How can you ensure that the the prefix and suffix of the eight bytes are F's with a single 32bit FF write?

Reply:
Hi, we looked closer, this function truncates how much it writes to not exceed word boundaries, that is why it is safe to only FF four bytes.

Perhaps there could be some documentation above the prototype explaining what the function is trying to do (as opposed to how it does it).

Reply:
Sure, I can add some docs about when it's called. How about:

```C
/**
 * Copies unaligned data into a word (u32_t).
 *
 * This function is used when either @p src or @p dst is an non-word aligned pointer.
 *
 * If the length exceeds the number of remaining in the word, the data is truncated.
 * Example: Attempting to copy 4 bytes into `0x0000 0001` will only copy the first
 * 3 bytes of the input data and return 3.
 *
 * @param[in,out] word_dst Pointer to 32-bit word appropriately filled with as much of
 *                         @p src as possible.
 * @param[in]     dst      Destination address pointer (flash). Used to calcuate
 *                         offset into word_dst. This is where the data should
 *                         eventually be copied into.
 * @param[in]     src      Source data pointer where data is copied from.
 * @param[in]     len      Length of the data pointed to by src.
 *
 *
 * @returns the number of bytes to copy.
 */
```

Reply:
LGTM

Reply:
`@returns the number of bytes to copy.`

Should be 

`@returns the number of bytes that were copied.`

Reply:
> Pointer to 32-bit word appropriately filled with as much of
>  *                         @p src as possible.


This sounds like it is already filled. Should it instead be saying 

> Pointer to 32-bit word to be appropriately filled with as much of
>  *                         @p src as possible.

Reply:
I can change the last one to `@returns the number of bytes copied into @p word_dst.` for clarity.
The second:
```
 * @param[in,out] word_dst Pointer to 32-bit word that will be appropriately
 *                         filled with as much of data from @p src as possible.
```